packages/origin.js/src/resources/purchases.js

```diff
@@ -1,48 +1,57 @@
-class Purchases {
-    constructor({ contractService, ipfsService }) {
-      this.contractService = contractService
-      this.ipfsService = ipfsService
-    }
+import ResourceBase from"../ResourceBase"
 
-    async contractFn(address, functionName, args=[], value=0){
-        const purchaseContract = this.contractService.purchaseContract
-        const purchase = await purchaseContract.at(address)
-        const account = await this.contractService.currentAccount()
-        args.push({ from: account, value: value })
-        return await purchase[functionName].apply(purchase, args)
-    }
+const _STAGES_TO_NUMBER = {
+  awaiting_payment: 0,
+  shipping_pending: 1,
+  buyer_pending: 2,
+  seller_pending: 3,
+  in_dispute: 4,
+  review_period: 5,
+  complete: 6
+}
+const _NUMBERS_TO_STAGE = {}
 
-    async get(address){
-        const contractData = await this.contractFn(
-            address,
-            'data'
-        )
-        return {
-            address: address,
-            stage: contractData[0], // perhaps return a string
-            listingAddress: contractData[1],
-            buyerAddress: contractData[2],
-            created: contractData[3],
-            buyerTimout: contractData[4]
-        }
-    }
 
-    async pay(address, amountWei){
-        return await this.contractFn(
-            address,
-            'pay',
-            [],
-            amountWei
-        )
-    }
+class Purchases extends ResourceBase{
+  constructor({ contractService, ipfsService }) {
+    super({ contractService, ipfsService })
+    
+    this.contractDefinition = this.contractService.purchaseContract
 
-    async buyerConfirmReceipt(address){
-        return await this.contractFn( address, 'buyerConfirmReceipt')
-    }
+    this.STAGES = {}
```

Comment:
@DanielVF I still don't understand why we want to export the stage definitions as part of the API. I've never used an API that exported a static definition of something. Can you provide an example of why this needs to be part of the exported object?

Reply:
1) IDE's and TypeScript can autocomplete (if I actually explicitly listed the stages, rather than auto generating - which I may do later). This really helps those folks.

2) It gives library users a nice code-accessible list of all valid purchase stages.

3) It lets IDE's and other javascript tools to reason about all the places that are actually working with stages or to make automated refactoring.

4) Catches typos. Typos go straight to undefined.

5) Not applicable to this exact code, but in other situations, it allows a library author to swap out values without affecting downstream code.

Reply:
1. Autocomplete is nice but it's just an editor-specific convenience. I don't think APIs should cater to editors like that. This point seems like a nice side benefit rather not a reason in and of itself.

2. Same as the first point.

3. I totally agree about defining constants this way in order to catch typos. Definitely a good pattern to follow. However, I think it's the consumer's responsibility to do watch out for typos in its own codebase. The API should be agnostic about how its code will be used. (Providing it in the API implies that the client will be referencing our defined constants rather than just using literals - now we're making assumptions about how are API is used externally).

5. This smells like a bad pattern. Yes that is technically possible; however, a well-designed API should not be changing these definitions except during major releases. Changing definitions should be a breaking change, and exporting a list of constant values that can actually change will encourage people to change these definitions in minor versions, with no mention of a breaking change. Now if someone happens to screw up and reference a literal value rather than using the exported list to make the conversion, their code doesn't work.

I'd be much more willing to accept this pattern if I could see an example of a respected code base using this pattern. I haven't seen it yet and I'm concerned it's not a good design pattern.

I hope I not blowing a small thing out of proportion, but I think it's an important question in terms of architecture design.

@wanderingstan Any thoughts? (To summarize, I am concerned that it's bad practice to be exporting constants in our API, especially as attributes on an instantiated class.)

Reply:
@DanielVF Those were vague because I haven't thought them through all the way. Wasn't a very good argument. I think to me it comes down to this being a different pattern than any APIs I've ever worked with. I want to stick with conventions.

For example, Stripe's API has similar attributes that can be one of multiple values. But they don't provide any sort of object or array or whatever that lets you fetch the list of possible values in the code. They simply list them out in the docs.

This is just one example, took less than a minute to find.

https://stripe.com/docs/api#payout_failures

How do you feel about following Stripe's API design as a guideline?

Reply:
Yes, exactly. I'm planning on always giving users of the api stages in string form, and I'm planning on always accepting stages from a user in string form.

Reply:
Since it might not have been clear - the API as it currently is only lets the API consumer see stages-as-strings. The consumer never knows about stages-as-numbers.